### DISCUSSION TITLE
Conditional compilation

### DISCUSSION THREAD

**Discussion by MartinSchwarzbauer** [mesonbuild/meson#12813] [Q&A]
Hi *! 

We are trying switching our buildsystem to "meson". We have different "platforms" where we wanna compile different source files!

E.g.: project(common)

```
if ( environment.get('platform') == 'a' )
  sources += "source_a.cpp"
endif

if ( environment.get('platform') == 'b' )
  sources += "source_b.cpp"
endif
```

My idea was, to decide this using an environment variable! -> but i have not found a way to access (get) environment variable! Set/Append is supported - why no 'get'?

I know i could use "options" ... but then i have to define an options file in each of our projects (about 40 projects). Using environment variable would we much easier (also in CLION IDE environment files are supported)

I hope i've described my "problem" so that my intention is clear!

Thank you very much,
Martin

**Comment by eli-schwartz** [comment]
Environment variables are just options, except that meson cannot track the persistent state of the options. If you check an environment variable and then edit meson.build and `ninja` runs `meson --internal regenerate`, then how do you know the environment variable is still set?

For some environment variables, like `$CC`, these are uplifted into persistent toolchains. This doesn't work for arbitrary environment variables, however.

**Comment by eli-schwartz** [comment]
It's possible to check an environment variable regardless, using

```meson
is_set = run_command('check_environment.py', 'VARNAME', check: true).stdout().strip() == 'set'
platform = run_command('check_environment_value.py', 'VARNAME', check: true).stdout().strip()
```
 or similar. This is stepping outside of the meson DSL, which means it depends on you guaranteeing the environment variable is always set when you need it to be.

**Comment by MartinSchwarzbauer** [comment]
Ok thank you for your quick reply! 
I understand, why it doesn't make sense to use environment variables!

But what's the way to do using meson?
In our SDK we set the variables (PATH, CC, and so on) =>  so I source this file before building! How to handle the "platform" specific stuff using meson?

Do i really have to add a meson.option file to each project?
When a new platform is added - i have to ajust ALL options in every project?

**Reply by eli-schwartz** [reply]
You shouldn't need to adjust all options in every project, just the ones in the projects where you're adding additional source files and editing the if/else branches.

On the other hand, if the option is just a string option accepting the name of a platform, you won't have to adjust it, ever. Because any value is valid as far as `meson.options` is concerned, and e.g. meson.build will check if that platform exists (has a corresponding source code file?) or error out otherwise.

**Comment by MartinSchwarzbauer** [comment]
Thank your for answer! 

Using meson 1.3.0 we are able to use

```
if compiler.has_define('MY_DEFINE')
    platform = 'my_platform'
elif ...

```

So this solved our problems!
